Add QRBills.create_esr_reference

kalsan opened the pull request:
Hello and thank you for your much appreciated work on this gem! It's a true gem indeed.

While creditor references (RF...) are very easy to generate with `QRBills.create_creditor_reference`, I've had quite some struggle to generate valid references for bills that support QR style IBANs. As it seems that I've finally managed to generate valid references (with an ESR-style modulo-10 check-digit), I thought that it might be a good idea to contribute the solution back to qr-bills, as it is a feature that is yet missing, in my opinion.

This pull request adds `QRBills.create_esr_reference` to `qr-bills`. The solution is based on http://sahits.ch/blog/blog/2007/11/08/uberprufen-esr-referenz-nummer/. Feel free to change anything you like :-)

Best,
Kalsan

kalsan committed:
Add QRBills.create_esr_reference

damoiser commented:
better later than never :-) sorry for the delay answering here and thanks a lot for your contribution.

Just a question, is ESR reference still compliant to QR bills? Reading quickly the six page I found that ESR got decommissioned in favor of the QR Bill IBAN

> The QR-IBAN replaces the ESR subscriber number.
https://www.six-group.com/en/blog/qr-bill-standing-orders.html

I just try to be more compliant as possible with the standards of QR Bills.

damoiser committed:
bump ruby version

damoiser committed:
bump gem version

damoiser committed:
add more error handling and tests

damoiser commented:
Hi @kalsan - thanks for the info, I slightly adjusted, bumped the ruby version and added some tests. Please check if is ok for you as well, after that. Will then merge it and publish the new version. Thanks again for contributing!

damoiser committed:
up license

damoiser commented:
Hi @kalsan - yes sure! feel free to commit as well the ``qr_iban?`` helper, however I would ask you to use another name for this function, maybe "qr_esr_iban?", as "qr_iban" can be interpreted as standard iban but this follow another validation pattern (as quickly checked here: https://stackoverflow.com/questions/21928083/iban-validation-check). 

Thanks!

kalsan committed:
Rename esr creditor reference method, add QRBills.iban_type and improve documentation.

kalsan commented:
@damoiser I've added an appropriate method and performed some slight refactoring for better consistency. Further, I've added a line to the README hinting to the comment explaining the differences.

damoiser commented:
Hi @kalsan - thank you for the commits and for the patience. Just published the new gem, available under the version v1.0.11: https://rubygems.org/gems/qr-bills/versions/1.0.11

Thanks again!
